examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py

```diff
@@ -0,0 +1,590 @@
+    lst_of_layers = list(l._flatten_layers())
+
+    if len(lst_of_layers) == 1:  # "leaves of the model"
+        if "A_" in l.name or "B_" in l.name:
+            l.trainable = True
+        else:
+            l.trainable = False
+
+"""
+Print the model's summary and see if the number of non-trainable parameters and
+total parameters are correct.
+
+In a previous section, we had calculated the number of parameters associated with
+the LoRA layers to be 6,144. The total trainable parameters in the model should
+be `num_layers * (query, value) * 6,144 = 12 * 2 * 6,144 = 147,456`. The
+number of non-trainable parameters should be the same as the total number of
+parameters in the original GPT-2 model, which is `124,439,808`.
+"""
+
+lora_model.summary()
+
+"""
+### Fine-tune LoRA GPT-2
+
+Now that we have hacked and verified the LoRA GPT-2 model, let's train it!
+Remember to subtract the current memory usage from the total memory usage to
+get the correct value for LoRA.
+"""
+
+gpu_memory_callback = GPUMemoryCallback(
+    target_batches=[5, 10, 25, 50, 100, 150, 200, 300, 400, 500],
+    subtract_value=current_memory_usage["current"],
+    print_stats=True,
+)
+
+# Linearly decaying learning rate.
+learning_rate = keras.optimizers.schedules.PolynomialDecay(
+    5e-5,
+    decay_steps=train_ds.cardinality() * EPOCHS,
+    end_learning_rate=0.0,
+)
+optimizer = keras.optimizers.Adam(learning_rate)
+loss = keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+
+lora_model.compile(
+    optimizer=optimizer,
+    loss=loss,
+    weighted_metrics=["accuracy"],
+)
+
+lora_model.fit(
+    train_ds,
+    epochs=EPOCHS,
+    callbacks=[gpu_memory_callback],
+)
+lora_model_memory_usage = gpu_memory_callback.memory_usage
+lora_model_memory_labels = gpu_memory_callback.labels
+
+"""
+And we are done fine-tuning the model! Before we generate text, let's compare
+the training time and memory usage of the two models. The training time of GPT-2
+on a 16 GB Tesla T4 (Colab) is 7 minutes, and for LoRA, it is 5 minutes, a 30%
+decrease. The memory usage of LoRA GPT-2 is roughly 3 times less than GPT-2.
+"""
+
+plt.plot(gpt2_lm_memory_labels, gpt2_lm_memory_usage, label="GPT-2")
+plt.plot(lora_model_memory_labels, lora_model_memory_usage, label="LoRA GPT-2")
+
+plt.xticks(rotation=90)
+
+plt.xlabel("Time")
+plt.ylabel("GPU Memory Usage")
+
+plt.title("GPU Memory Usage (in GB)")
+plt.legend()
+plt.show()
+
+"""
+### Merge weights and generate text!
+
+One of the biggest advantages of LoRA over other adapter methods is that it
+does not incur any additional inference latency. Let's understand why.
+
+Recall our LoRA equation: `output = W0x + b0 + BAx`. We can rewrite this as:
+`output = = Wx + b0 = (W0 + BA)x + b0`, where `W = W0 + BA`. This means that if
+we merge the weights of the original model and the adapter, we will be essentially
+doing the same computation as the original model!
+"""
+
+# Freeze the whole model.
+lora_model.trainable = False
```

Comment:
why is this needed? i don't see any gradient descent in the rest of the guide

Reply:
Isn't it useful for demonstration purposes?

Reply:
Isn't it useful for demonstration purposes? Can remove if it feel redundant

Reply:
I guess to me I think of `trainable` as only pertaining to `fit` calls. So it threw me for a bit of a loop here and I was wondering if I missed something below. You could add a comment above the generate call maybe `# Freezing weights not necessary during generation, no weights are updated.`